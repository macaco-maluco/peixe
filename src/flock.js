import { Vector3 } from 'three'

const NEIGHBOUR_RADIUS = 100
const SEPARATION_WEIGHT = 3.5
const ALIGNMENT_WEIGHT = 0.2
const COHESION_WEIGHT = 4

export const MAX_SPEED = 0.2

export default function flock(boid, neighbours) {
  const separation = separate(boid, neighbours).multiplyScalar(SEPARATION_WEIGHT)
  const alignment = align(boid, neighbours).multiplyScalar(ALIGNMENT_WEIGHT)
  const cohesion = cohere(boid, neighbours).multiplyScalar(COHESION_WEIGHT)

  const acceleration = separation
    .add(alignment)
    .add(cohesion)
    // Slow the acceleration down by a big factor
    .divideScalar(200)

  return { acceleration, separation, alignment, cohesion }
}

function cohere(boid, neighbours) {
  const flockMembers = neighbours.filter(neighbour => {
    const distance = boid.position.distanceTo(neighbour.position)
    return distance > 0 && distance < NEIGHBOUR_RADIUS
  })

  if (flockMembers.length === 0) {
    return new Vector3(0, 0, 0)
  }

  const centerOfMass = flockMembers
    .reduce((acc, neighbour) => acc.add(neighbour.position), new Vector3())
    .divideScalar(flockMembers.length)
    .normalize()

  const cohesion = centerOfMass
    .clone()
    .sub(boid.position)
    .normalize()

  return cohesion
}

function align(boid, neighbours) {
  const flockMembers = neighbours.filter(neighbour => {
    const distance = boid.position.distanceTo(neighbour.position)
    return distance > 0 && distance < NEIGHBOUR_RADIUS
  })

  if (flockMembers.length === 0) {
    return new Vector3(0, 0, 0)
  }

  return flockMembers
    .reduce((acc, neighbour) => acc.add(neighbour.velocity), new Vector3())
    .divideScalar(flockMembers.length)
    .normalize()
}

function separate(boid, neighbours) {
  const flockMembers = neighbours.filter(neighbour => {
    const distance = boid.position.distanceTo(neighbour.position)
    return distance > 0 && distance < NEIGHBOUR_RADIUS
  })

  if (flockMembers.length === 0) {
    return new Vector3(0, 0, 0)
  }

  return (
    flockMembers
      .reduce((acc, neighbour) => {
        return boid.position
          .clone()
          .sub(neighbour.position)
          .add(acc)
      }, new Vector3())
      .divideScalar(flockMembers.length)
      // .multiplyScalar(-1)
      .normalize()
  )
}
