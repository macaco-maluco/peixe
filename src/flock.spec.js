import { Vector3 } from 'three'
import flock from './flock'

it('should flock', () => {
  const boid = { position: new Vector3(0, 0, 0), velocity: new Vector3(0, 0, 0) }

  const neighbours = [
    { position: new Vector3(0.2, 0.1, 0), velocity: new Vector3(0, 0, 0) },
    { position: new Vector3(1, 0, 0), velocity: new Vector3(0, 0, 0) },
  ]

  const acceleration = flock(boid, neighbours)

  expect(acceleration).toMatchSnapshot()
})
