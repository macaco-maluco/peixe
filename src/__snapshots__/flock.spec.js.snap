// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should flock 1`] = `
Object {
  "acceleration": Vector3 {
    "x": 4.440892098500626e-16,
    "y": 2.7755575615628914e-17,
    "z": 0,
  },
  "alignment": Vector3 {
    "x": 0,
    "y": 0,
    "z": 0,
  },
  "cohesion": Vector3 {
    "x": 1.9930915164897596,
    "y": 0.16609095970747997,
    "z": 0,
  },
  "separation": Vector3 {
    "x": 4.440892098500626e-16,
    "y": 2.7755575615628914e-17,
    "z": 0,
  },
}
`;
